File: trees_and_graphs/graph_coloring/color_graph.js
class GraphNode {
  constructor(label) {
    this.label = label;
    this.neighbors = new Set();
    this.color = null;
  }
}

function colorGraph(graph, colors) {


  // Create a valid coloring for the graph
  // each item in the graph is a GraphNode, with a label, set of neighbor nodes, and unset color
  // this function will give a valid color to each node
  // each node can have maximum degree of D, and thus only D + 1 colors can be used
  // Iterate thru the graph array, and with each node, check its neighbors (which should be at most D neighbors)
  // then iterate thru the colors array and check that the current color is not in the set of neighbor's colors, and use it
  for (let node of graph) {
    let neighbors = node.neighbors;
    
    if (neighbors.has(node)) {
      throw new ValueError("Cannot have self as neighbor");
    }
    // I can improve the solution here with a Set instead of array
    // Need to build the set my iterating thru neighbors set and building a temp D number set of colors
    let illegalNeighborColors = new Set();
    neighbors.forEach(neighbor => {
      if (neighbor.color && !illegalNeighborColors.has(neighbor.color)) {
        illegalNeighborColors.add(neighbor.color);
      }
    });
    // let neighborColors = [...neighbors].map( (neighborNode) => {
    //   return neighborNode.color;
    // })

    // only need to check D + 1 colors, where D is the number of neighbors
    const maxColors = neighbors.size + 1;
    for (let i = 0; i < maxColors; i++) {
      const currentColor = colors[i];
      if (illegalNeighborColors.has(currentColor)) {
        continue;
      } else {
        node.color = currentColor;
        break;
      }
    }

    // console.log("Neighbor colors below for node", node.label)
    // console.log(neighborColors);

  }
}

const colors = ['red', 'green', 'blue', 'orange', 'yellow', 'white'];
let graph = [];
{
  const nodeA = new GraphNode('A');
  const nodeB = new GraphNode('B');
  const nodeC = new GraphNode('C');
  const nodeD = new GraphNode('D');
  nodeA.neighbors.add(nodeB);
  nodeB.neighbors.add(nodeA);
  nodeB.neighbors.add(nodeC);
  nodeC.neighbors.add(nodeB);
  nodeC.neighbors.add(nodeD);
  nodeD.neighbors.add(nodeC);
  graph = [nodeA, nodeB, nodeC, nodeD];
}
colorGraph(graph, colors);
console.log(graph);

File: trees_and_graphs/graph_coloring/color_graph.py
class GraphNode:

  def __init__(self, label):
    self.label = label
    self.neighbors = set()
    self.color = None


def color_graph(graph, colors):

  # Create a valid coloring for the graph
  # Iterate thru the N nodes
  # for each node, look at it's nearest neighbors
  # the total number of neighbors we would look at is 2 * M where M is the edges connecting two neighbors
  # with the neighbors, we will get the Set of colors that we cannot apply to our current node
  # then check our array of D colors, and if the color is not in the set, we can apply it to our current node

  # we have N Nodes
  for graph_node in graph:
    # at most D neighbors
    # over the span of the loop, we will have a total of 2M neighbors, where M is the edges between nodes
    neighbors = graph_node.neighbors

    # make sure graph node is not its own neighbor to prevent any loops
    if graph_node in neighbors:
      raise ValueError("A node cannot be its own neighbor")

    illegal_colors = set()
    for neighbor in neighbors:
      if neighbor.color:
        illegal_colors.add(neighbor.color)

    # make sure a color we choose is not in the set of illegal colors  
    for color in colors:
      if color not in illegal_colors:
        graph_node.color = color
        break
  
  return 0

